pdf-text-api/src/main/resources/xpdf/linux/doc/pdftotext.1
.\" Copyright 1997-2022 Glyph & Cog, LLC
.TH pdftotext 1 "18 Apr 2022"
.SH NAME
pdftotext \- Portable Document Format (PDF) to text converter
(version 4.04)
.SH SYNOPSIS
.B pdftotext
[options]
.RI [ PDF-file
.RI [ text-file ]]
.SH DESCRIPTION
.B Pdftotext
converts Portable Document Format (PDF) files to plain text.
.PP
Pdftotext reads the PDF file,
.IR PDF-file ,
and writes a text file,
.IR text-file .
If
.I text-file
is not specified, pdftotext converts
.I file.pdf
to
.IR file.txt .
If 
.I text-file
is \'-', the text is sent to stdout.
.SH CONFIGURATION FILE
Pdftotext reads a configuration file at startup.  It first tries to
find the user's private config file, ~/.xpdfrc.  If that doesn't
exist, it looks for a system-wide config file, typically /etc/xpdfrc
(but this location can be changed when pdftotext is built).  See the
.BR xpdfrc (5)
man page for details.
.SH OPTIONS
Many of the following options can be set with configuration file
commands.  These are listed in square brackets with the description of
the corresponding command line option.
.TP
.BI \-f " number"
Specifies the first page to convert.
.TP
.BI \-l " number"
Specifies the last page to convert.
.TP
.B \-layout
Maintain (as best as possible) the original physical layout of the
text.  The default is to \'undo' physical layout (columns,
hyphenation, etc.) and output the text in reading order.  If the
.B \-fixed
option is given, character spacing within each line will be determined
by the specified character pitch.
.TP
.B \-simple
Similar to
.BR \-layout ,
but optimized for simple one-column pages.  This mode will do a better
job of maintaining horizontal spacing, but it will only work properly
with a single column of text.
.TP
.B \-simple2
Similar to
.BR \-simple ,
but handles slightly rotated text (e.g., OCR output) better.  Only works
for pages with a single column of text.
.TP
.B \-table
Table mode is similar to physical layout mode, but optimized for
tabular data, with the goal of keeping rows and columns aligned (at
the expense of inserting extra whitespace).  If the
.B \-fixed
option is given, character spacing within each line will be determined
by the specified character pitch.
.TP
.B \-lineprinter
Line printer mode uses a strict fixed-character-pitch and -height
layout.  That is, the page is broken into a grid, and characters are
placed into that grid.  If the grid spacing is too small for the
actual characters, the result is extra whitespace.  If the grid
spacing is too large, the result is missing whitespace.  The grid
spacing can be specified using the
.B \-fixed
and
.B \-linespacing
options.
If one or both are not given on the command line, pdftotext will
attempt to compute appropriate value(s).
.TP
.B \-raw
Keep the text in content stream order.  Depending on how the PDF file
was generated, this may or may not be useful.
.TP
.BI \-fixed " number"
Specify the character pitch (character width), in points, for physical
layout, table, or line printer mode.  This is ignored in all other
modes.
.TP
.BI \-linespacing " number"
Specify the line spacing, in points, for line printer mode.  This is
ignored in all other modes.
.TP
.B \-clip
Text which is hidden because of clipping is removed before doing
layout, and then added back in.  This can be helpful for tables where
clipped (invisible) text would overlap the next column.
.TP
.B \-nodiag
Diagonal text, i.e., text that is not close to one of the 0, 90, 180,
or 270 degree axes, is discarded.  This is useful to skip watermarks
drawn on top of body text, etc.
.TP
.BI \-enc " encoding-name"
Sets the encoding to use for text output.  The
.I encoding\-name
must be defined with the unicodeMap command (see
.BR xpdfrc (5)).
The encoding name is case-sensitive.  This defaults to "Latin1" (which
is a built-in encoding).
.RB "[config file: " textEncoding ]
.TP
.BI \-eol " unix | dos | mac"
Sets the end-of-line convention to use for text output.
.RB "[config file: " textEOL ]
.TP
.B \-nopgbrk
Don't insert a page breaks (form feed character) at the end of each
page.
.RB "[config file: " textPageBreaks ]
.TP
.B \-bom
Insert a Unicode byte order marker (BOM) at the start of the text
output.
.TP
.BI \-marginl " number"
Specifies the left margin, in points.  Text in the left margin (i.e.,
within that many points of the left edge of the page) is discarded.
The default value is zero.
.TP
.BI \-marginr " number"
Specifies the right margin, in points.  Text in the right margin
(i.e., within that many points of the right edge of the page) is
discarded.  The default value is zero.
.TP
.BI \-margint " number"
Specifies the top margin, in points.  Text in the top margin (i.e.,
within that many points of the top edge of the page) is discarded.
The default value is zero.
.TP
.BI \-marginb " number"
Specifies the bottom margin, in points.  Text in the bottom margin
(i.e., within that many points of the bottom edge of the page) is
discarded.  The default value is zero.
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.B \-verbose
Print a status message (to stdout) before processing each page.
.RB "[config file: " printStatusInfo ]
.TP
.B \-q
Don't print any messages or errors.
.RB "[config file: " errQuiet ]
.TP
.BI \-cfg " config-file"
Read
.I config-file
in place of ~/.xpdfrc or the system-wide config file.
.TP
.B \-listencodings
List all available text output encodings, then exit.
.TP
.B \-v
Print copyright and version information, then exit.
.TP
.B \-h
Print usage information, then exit.
.RB ( \-help
and
.B \-\-help
are equivalent.)
.SH BUGS
Some PDF files contain fonts whose encodings have been mangled beyond
recognition.  There is no way (short of OCR) to extract text from
these files.
.SH EXIT CODES
The Xpdf tools use the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
2
Error opening an output file.
.TP
3
Error related to PDF permissions.
.TP
99
Other error.
.SH AUTHOR
The pdftotext software and documentation are copyright 1996-2022 Glyph
& Cog, LLC.
.SH "SEE ALSO"
.BR xpdf (1),
.BR pdftops (1),
.BR pdftohtml (1),
.BR pdfinfo (1),
.BR pdffonts (1),
.BR pdfdetach (1),
.BR pdftoppm (1),
.BR pdftopng (1),
.BR pdfimages (1),
.BR xpdfrc (5)
.br
.B http://www.xpdfreader.com/


pdf-to-text/src/main/java/io/xpdfutils/pdftotext/PdfToText.java
package io.xpdfutils.pdftotext;

import io.xpdfutils.common.*;
import lombok.val;
import org.apache.commons.io.FileUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

//todo: in the future, extend Callable so that users of sdk can run asynchronously if they would prefer
public class PdfToText implements XpdfUtility<PdfToTextRequest, PdfToTextResponse> {
    private final String binCommand;
    //todo: are threads managed correctly with process here?
    // should new singleton be declared, or retrieved
    // should you properly shutdown when done?
//    private final ExecutorService executorService = Executors.newSingleThreadExecutor();

    public PdfToText() {
        //todo: add unit tests to ensure all os/bit value combos have resource
        val os = XpdfOperatingSystem.get();

        //todo: whats the best way to distribute resources?
        //      this solution copies binaries resource from inside jar to a directory outside of jar which accessible to client OS...
        //      but is there a better way? this solution feels dirty
        //      should we request client download the binaries themself, and configure path?
        //todo: also, there is no way for the client to verify that we are including the authentic xpdf binaries in this solution...
        //      how can you package the binaries with this solution in a credible way?
        //      maybe some way to incorporate the pgp key provided on xpdf website into build/distribution process? https://www.xpdfreader.com/download.html
        //todo: instead of copying resource every time an instance is created, a check should first be performed to make sure it doesnt already exist
        //      this same check should also be done in the process() method
        //      maybe move some of this code into common..
        val binName = "windows".equals(os.getOperatingSystem()) ? "pdftotext.exe" : "pdftotext";
        val binResourceStream = getClass().getClassLoader().getResourceAsStream("xpdf/%s/%s/%s".formatted(os.getOperatingSystem(), os.getBit(), binName));
        if (binResourceStream == null)
            throw new XpdfRuntimeException("Unable to locate pdftotext binaries");

        try {
            // copy bin resource to directory accessible to client OS
            val binAccessiblePath = Paths.get(System.getProperty("java.io.tmpdir"), "xpdf", binName);
            FileUtils.copyInputStreamToFile(binResourceStream, binAccessiblePath.toFile());
            binCommand = binAccessiblePath.toFile().getCanonicalPath();
        } catch (Exception e) {
            throw new XpdfRuntimeException("Unable to copy pdftotext binaries to directory accessible by OS");
        }

        // old approach
//        val exeDirectory = getClass().getClassLoader().getResource("xpdf/%s/%s".formatted(os.getOperatingSystem(), os.getBit()));
        //        if (exeDirectory == null)
//            throw new XpdfRuntimeException("Unable to locate pdftotext binaries");
//        try {
//            exeCommand = Paths.get(Paths.get(exeDirectory.toURI()).toFile().getCanonicalPath(),"pdftotext").toFile().getCanonicalPath();
//        } catch (Exception e) {
//            throw new XpdfRuntimeException("Unable to get path of pdftotext binaries");
//        }
    }

    //todo: is "process" really the most friendly name for this?
    // maybe you should just drop the interface and simplify this
    //todo: add @NotNull to public method parameters
    @Override
    public PdfToTextResponse process(PdfToTextRequest request) throws XpdfProcessingException, XpdfValidationException {
        val executorService = Executors.newSingleThreadExecutor();

        try {
            // validate request
            validate(request);

            // get commands
            //todo: how can you be sure user is not injecting malicious arguments into file path..?
            // need to verify args better..
            val commands = new ArrayList<String>();
            commands.add(binCommand);
            commands.addAll(getCommandOptions(request.getOptions()));
            commands.add(request.getPdfFile().getCanonicalPath());
            commands.add(request.getTxtFile().getCanonicalPath());

            // process commands
            val processBuilder = new ProcessBuilder(commands.toArray(new String[0]));
            val process = processBuilder.start();

            //todo: log output
            //todo: should you do this after waiting for process to finish, so that processing of pdf does not hang on waiting for std out?
            // or makes sense as is?
            val standardOutput = executorService.submit(new ReadInputStreamTask(process.getInputStream())).get();
            val errorOutput = executorService.submit(new ReadInputStreamTask(process.getErrorStream())).get();

            //todo: configurable timeout?
            val exitCode = process.waitFor();

            // handle process result
            if (exitCode == 0) {
                return PdfToTextResponse.builder()
                        .pdfText(getPdfText(request))
                        .stdOut(standardOutput)
                        .build();
            } else {
                final String message;
                switch (exitCode) {
                    case 1 -> message = "Error opening the PDF file";
                    case 2 -> message = "Error opening the output file";
                    case 3 -> message = "Error related to PDF permissions";
                    case 99 -> message = "Other Xpdf error";
                    default -> message = "Unknown Xpdf error";
                }

                throw new XpdfProcessingException(standardOutput, errorOutput, message);
            }
        } catch (XpdfProcessingException | XpdfValidationException | XpdfRuntimeException e) {
            throw e;
        } catch (Exception e) {
            //todo: should we be throwing checked or unchecked exceptions?
            throw new XpdfRuntimeException(e);
        } finally {
            executorService.shutdown();
        }
    }

    protected void validate(PdfToTextRequest request) throws XpdfValidationException {
        // verify files
        if (!request.getPdfFile().exists())
            throw new XpdfValidationException("PdfFile does not exist");

        try {
            request.getTxtFile().getCanonicalPath();
        } catch (IOException e) {
            throw new XpdfValidationException("TxtFile does not have a valid path");
        }

        // verify options
        val pageStart = request.getOptions().getPageStart();
        val pageEnd = request.getOptions().getPageEnd();
        if (pageStart != null && pageStart < 0)
            throw new XpdfValidationException("PageStart cannot be less than zero");
        if (pageEnd != null && pageEnd < 0)
            throw new XpdfValidationException("PageEnd cannot be less than zero");
        if (pageStart != null && pageEnd != null && pageStart > pageEnd)
            throw new XpdfValidationException("PageStart must come before PageEnd");

        //todo: what other validation would be helpful?
    }

    protected List<String> getCommandOptions(PdfToTextRequest.Options options) {
        if (options == null)
            return List.of();

        val args = new ArrayList<String>();

        //todo: add by default? or add when in debug? or add as option? or make configurable?
//        args.add("-verbose");

        val pageStart = options.getPageStart();
        if (pageStart != null)
            args.addAll(List.of("-f", pageStart.toString()));

        val pageEnd = options.getPageEnd();
        if (pageEnd != null)
            args.addAll(List.of("-l", pageEnd.toString()));

        val format = options.getFormat();
        if (format != null) {
            switch (format) {
                case RAW -> args.add("-raw");
                case SIMPLE -> args.add("-simple");
                case TABLE -> args.add("-table");
                case LAYOUT -> args.add("-layout");
                case LINE_PRINTER -> args.add("-lineprinter");
                default -> throw new XpdfRuntimeException("Format case %s is missing from command options switch statement".formatted(format.name()));
            }
        }

        val encoding = options.getEncoding();
        if (encoding != null) {
            switch (encoding) {
                case ASCII_7 -> args.addAll(List.of("-enc", "ASCII7"));
                case LATIN_1 ->args.addAll(List.of("-enc", "Latin1"));
                case SYMBOL -> args.addAll(List.of("-enc", "Symbol"));
                case UCS_2 -> args.addAll(List.of("-enc", "UCS-2"));
                case UTF_8 -> args.addAll(List.of("-enc", "UTF-8"));
                case ZAPF_DINGBATS -> args.addAll(List.of("-enc", "ZapfDingbats"));
                default -> throw new XpdfRuntimeException("Encoding case %s is missing from command options switch statement".formatted(encoding.name()));
            }
        }

        val endOfLine = options.getEndOfLine();
        if (endOfLine != null) {
            switch (endOfLine) {
                case DOS -> args.addAll(List.of("-eol", "dos"));
                case MAC -> args.addAll(List.of("-eol", "mac"));
                case UNIX -> args.addAll(List.of("-eol", "unix"));
                default -> throw new XpdfRuntimeException("EndOfLine case %s is missing from command options switch statement".formatted(endOfLine.name()));
            }
        }

        val pageBreakIncluded = options.isPageBreakIncluded();
        if (!pageBreakIncluded)
            args.add("-nopgbrk");

        val ownerPassword = options.getOwnerPassword();
        if (ownerPassword != null)
            args.addAll(List.of("-opw", "\"%s\"".formatted(ownerPassword)));

        val userPassword = options.getUserPassword();
        if (userPassword != null)
            args.addAll(List.of("-upw", "\"%s\"".formatted(userPassword)));

        return args;
    }

    protected String getPdfText(PdfToTextRequest request) throws IOException {
        //todo: need to figure out encoding stuff...
        // when you specify encoding to xpdf, that helps it understand how to read the document.
        // but how does it decide to encode the data into the text file? does it retain same specified encoding?
        // run some tests...
        return String.join("\n", Files.readAllLines(request.getTxtFile().toPath(), StandardCharsets.UTF_8));
    }
}
